### DISCUSSION TITLE
Combining Ops and Assets within the same dagster job

### DISCUSSION THREAD

**Discussion by IsaacChanXi** [dagster-io/dagster#19162] [Q&A] [answered]
Dagster documentation and guides show how jobs can be composed either by ops or by assets. I'm wondering if there's a solution whereby triggering the job triggers both assets and ops that are under that job.

A simple hypothetical use case is to execute a task, such as an API call, before the assets are run. I believe the API call should be an op, as it does not produce any data or object.

Would love to hear if anyone has suggestions to this!

P.S, I've come across graphs that can be used as an asset, but a simple API call should not be composed as a graph.

**Comment by OwenKephart** [comment, marked as answer]
Hi @IsaacChanXi -- could you elaborate more on the issue with making a graph that is (simple API call op) -> (asset producing op)? That seems like a reasonable solution, and graph-backed assets are the only supported way of combining op-based workflows into an asset-based job.

The other option would be to put the API call inside the body of the asset-producing op.

**Reply by IsaacChanXi** [reply]
Hi Owen, thanks for your reply. I thought that the API call should be an op since it's more of a "task", rather than an asset that produces some object. 

But you are right that an API call can be put in the body of an asset-producing op, that would make more sense. 

Thank you!
